lib/server.js

```diff
@@ -0,0 +1,205 @@
+import http from 'node:http'
+import net from 'node:net'
+import {EventEmitter} from 'node:events'
+import crypto from 'node:crypto'
+import url from 'node:url'
+
+import {Headers} from 'undici'
+
+import {SafeHttpClient, HttpError} from './safe-http-client.js'
+import {
+  DEFAULT_HEADERS,
+  REQUEST_HEADERS,
+  RESPONSE_HEADERS
+} from './constants.js'
+
+/**
+ * @typedef {Object} Config
+ * @property {string} HMACKey
+ * @property {string} serverName Server name used in Headers and Via checks
+ * @property {number|undefined} [maxSize] The max size in bytes per resource to download. Sends 500 if exceeded.
+ */
+
+/**
+ * @param {string} hex
+ */
+function hexDecode(hex) {
+  const bytes = []
+  for (let i = 0; i < hex.length; i += 2) {
+    bytes.push(parseInt(hex.substr(i, 2), 16))
+  }
+  return Buffer.from(bytes)
+}
+
+/**
+ * Filter incoming headers based on allowed headers.
+ * We always want to send Pascal-Case headers so we use the key from constants
+ * with the value from the incoming header.
+ * @param {Map<string, boolean>} allowedHeaders
+ * @returns {(incomingHeaders: Headers) => Record<string, string>}
+ * */
+function filterHeaders(allowedHeaders) {
+  return function (incomingHeaders) {
+    /** @type {Record<string, string>} */
+    const filteredHeaders = {}
+
+    for (const [key, bool] of allowedHeaders.entries()) {
+      if (incomingHeaders.has(key.toLowerCase()) && bool) {
+        // @ts-ignore
+        filteredHeaders[key] = incomingHeaders.get(key.toLowerCase())
+      }
```

Comment:
```suggestion
      const value = incomingHeaders.get(key.toLowerCase())

      if (value !== null && value !== undefined && bool) {
        filteredHeaders[key] = value
      }
```

Solve the type error, and don’t first check and then get, which is an antipattern: get and then check whether it’s defined or not.

Two Qs:
a) are the headers case insensitive? If so, that should likely be noted in `constant.js`. I’d also then say we shouldn’t case them there at all?
b) This also seems like the `bool` isn’t useful: `false` is treated the same as not being included in `allowedHeaders` at all. Perhaps sets are better?

Reply:
> a) are the headers case insensitive? If so, that should likely be noted in `constant.js`. I’d also then say we shouldn’t case them there at all?

According to the HTTP spec it doesn't matter. But it's definitely canonical to use Pascal-Case.

> This also seems like the `bool` isn’t useful: `false` is treated the same as not being included in `allowedHeaders` at all. Perhaps sets are better?

It's handy to explicitly indicate what is not allowed and when respecting forwarded headers are allowed, we could change the value to true based on some checks. But `Set` may be good too here yes.